src/check_lines.rs

```diff
@@ -190,6 +190,25 @@ pub fn check_lines(
         }
     }
 
+    // Enfore newlines at end of files. Don't check if noqa enabled.
+    if settings.enabled.contains(&CheckCode::W292) && !enforce_noqa {
```

Comment:
You can remove `!enforce_noqa`, which is used for something else. You _do_ need to check if the last line has a `noqa` for W292. I'm also happy to add that in a second pass since the logic can be a little tricky. But the gist of it is:

```rust
let lineno = lines.len() - 1;
let 
let noqa_lineno = noqa_line_for
    .get(lineno)
    .map(|lineno| lineno - 1)
    .unwrap_or(lineno);
let check = Check::new(
    CheckKind::NoNewLineAtEndOfFile,
    Range {
        location: Location::new(lines.len(), line.len() + 1),
        end_location: Location::new(lines.len(), line.len() + 1),
    },
);
match noqa {
    (Directive::All(_, _), matches) => {
        matches.push(check.kind.code().as_str());
    }
    (Directive::Codes(_, _, codes), matches) => {
        if codes.contains(&check.kind.code().as_str()) {
            matches.push(check.kind.code().as_str());
        } else {
            line_checks.push(check);
        }
    }
    (Directive::None, _) => line_checks.push(check),
}
```

(Or something like that...)

Reply:
Yea i noticed this from a check above. I think I was thinking of how I could do something similar without slapping in redundant code, and I didn't feel confident introducing a refactor. So I must of misunderstood `enforce_noqa`. I'll see if I can wrap my head around this.

Reply:
So that was pretty much it. Only a couple things seemed to be needed in addition:

1. Adding
```rs
let noqa = noqa_directives
    .entry(noqa_lineno)
    .or_insert_with(|| (noqa::extract_noqa_directive(lines[noqa_lineno]), vec![]));
```
which if I understand is noqa indicators potentially used and/or related to the check that we need to attempt to match on? It does feels a little black-magic-like at the moment, but I think at a high level I understand what's going on here. I think if I understand the Directives better it'll clear up.

2. `noqa_directives` was moved in the "Enforce that the noqa directive was actually used" piece. So I needed to move the check above it.

Reply:
Yeah the noqa stuff is a little confusing. There are two main pieces to it:

1. It's not always the case that a `noqa` for a given line of code comes at the end of that line. For example, with multi-line strings, we want the `noqa` to come after the string, but to apply to the _entire_ string. So we have to do some bookkeeping, to map "line in the code" to "line at which the noqa for that line should appear". (This is the `noqa_line_for` vector, which is really a map.)
2. We have to extract the `noqa` directive from each line, and we do that lazily, hence the `noqa_directives` map.